System/Console/ANSI/Codes.hs

```diff
@@ -0,0 +1,161 @@
+        FaintIntensity  -> 2
+        NormalIntensity -> 22
+    SetItalicized True  -> 3
+    SetItalicized False -> 23
+    SetUnderlining underlining -> case underlining of
+        SingleUnderline -> 4
+        DoubleUnderline -> 21
+        NoUnderline     -> 24
+    SetBlinkSpeed blink_speed -> case blink_speed of
+        SlowBlink   -> 5
+        RapidBlink  -> 6
+        NoBlink     -> 25
+    SetVisible False -> 8
+    SetVisible True  -> 28
+    SetSwapForegroundBackground True  -> 7
+    SetSwapForegroundBackground False -> 27
+    SetColor Foreground Dull color  -> 30 + colorToCode color
+    SetColor Foreground Vivid color -> 90 + colorToCode color
+    SetColor Background Dull color  -> 40 + colorToCode color
+    SetColor Background Vivid color -> 100 + colorToCode color
+
+cursorUpCode, cursorDownCode, cursorForwardCode, cursorBackwardCode :: Int -- ^ Number of lines or characters to move
+                                                                    -> String
+cursorUpCode n = csi [n] "A"
+cursorDownCode n = csi [n] "B"
+cursorForwardCode n = csi [n] "C"
+cursorBackwardCode n = csi [n] "D"
+
+cursorDownLineCode, cursorUpLineCode :: Int -- ^ Number of lines to move
+                                     -> String
+cursorDownLineCode n = csi [n] "E"
+cursorUpLineCode n = csi [n] "F"
+
+setCursorColumnCode :: Int -- ^ 0-based column to move to
+                    -> String
+setCursorColumnCode n = csi [n + 1] "G"
+
+setCursorPositionCode :: Int -- ^ 0-based row to move to
+                      -> Int -- ^ 0-based column to move to
+                      -> String
+setCursorPositionCode n m = csi [n + 1, m + 1] "H"
+
+clearFromCursorToScreenEndCode, clearFromCursorToScreenBeginningCode, clearScreenCode :: String
+clearFromCursorToLineEndCode, clearFromCursorToLineBeginningCode, clearLineCode :: String
+
+clearFromCursorToScreenEndCode = csi [0] "J"
+clearFromCursorToScreenBeginningCode = csi [1] "J"
+clearScreenCode = csi [2] "J"
+clearFromCursorToLineEndCode = csi [0] "K"
+clearFromCursorToLineBeginningCode = csi [1] "K"
+clearLineCode = csi [2] "K"
+
+scrollPageUpCode, scrollPageDownCode :: Int -- ^ Number of lines to scroll by
+                                     -> String
+scrollPageUpCode n = csi [n] "S"
+scrollPageDownCode n = csi [n] "T"
+
+setSGRCode :: [SGR] -- ^ Commands: these will typically be applied on top of the
+                    -- current console SGR mode. An empty list of commands is
+                    -- equivalent to the list @[Reset]@. Commands are applied
+                    -- left to right.
+           -> String
+setSGRCode sgrs = csi (map sgrToCode sgrs) "m"
+
+hideCursorCode, showCursorCode :: String
+hideCursorCode = csi [] "?25l"
+showCursorCode = csi [] "?25h"
+
+
+-- | XTerm control sequence to set the Icon Name and Window Title.
+setTitleCode :: String -- ^ New Icon Name and Window Title
+             -> String
+setTitleCode title = "\ESC]0;" ++ filter (/= '\007') title ++ "\007"
```

Comment:
They are somewhat important, but you are not supposed to "add" them by hand -- that would be annoying and error-prone. Instead, as I said, you should configure your editor to add them automatically. On UNIX, newline characters act as line terminators, not separators, so the last newline character simply ends the last line.